### PR TITLE
Adding dialogflow_cx_test_case resource

### PR DESCRIPTION
**Pull request by mmurakowski-verily**
Adds a a test_case resource. Fixes https://github.com/hashicorp/terraform-provider-google/issues/15744

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_dialogflow_cx_test_case`
```

### CODE REVIEW

**Comment by mmurakowski-verily** [comment]
It's part of the API, but I'm iffy about if it's worthwhile to have `whatever.displayName`: if you have both, you're just repeating yourself; but if you only have one, `terraform apply` will keep trying to remove the other after the resource is created.

**Comment by melinath** [comment]
In terms of "flattening" nested fields to something like:

```
triggered_intent = google_dialogflow_cx_intent.some_intent.id
current_page = google_dialogflow_cx_page.some_page.id
```

We don't allow that in the provider any more unless it's to match an existing implementation (for example, a sibling resource that was implemented that way), because there have been too many cases where other nested fields ended up being added later and causing confusion / bugs / extra work.

---

Regarding the display_name fields - if the API _requires_ them to always be equal to the `google_dialogflow_cx_intent.display_name` and `google_dialogflow_cx_page.display_name` fields, then it's probably fine to not add them to the resource for now (or make them output-only). Folks can just reference the value from the relevant resource directly. If there's a valid reason you're aware of that it would be different than the intent/page resource's display_name, then implementing it would be preferable.

**Comment by mmurakowski-verily** [comment]
> too many cases where other nested fields ended up being added later and causing confusion / bugs / extra work

Yeah, I figured; oh well.

I tested it and it looks like if you provide a `display_name` and `name` that don't match, `name` wins; it picks the Intent with the ID (`name`) you gave it, and uses that one's `display_name` as the `triggered_intent.display_name`. I made the `display_name` output only and added a comment so hopefully nobody gets confused in the future.

File: mmv1/products/dialogflowcx/TestCase.yaml [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+  pre_read: templates/terraform/pre_create/dialogflow_set_location.go.erb
+examples:
+  - !ruby/object:Provider::Terraform::Examples
+    name: 'dialogflowcx_test_case_full'
+    primary_resource_id: 'basic_test_case'
+    vars:
+      agent_name: 'dialogflowcx-agent'
+skip_sweeper: true
+id_format: '{{parent}}/testCases/{{name}}'
+import_format: ['{{parent}}/testCases/{{name}}']
```
Comment (melinath) [inline mmv1/products/dialogflowcx/TestCase.yaml:42]: If the only custom import logic is handling slashes in the parent name, it should be possible to use this import_format instead.

```suggestion
import_format: ['{{%parent}}/testCases/{{name}}']
```

File: mmv1/products/dialogflowcx/TestCase.yaml [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+  You can use the built-in test feature to uncover bugs and prevent regressions. A test execution verifies that agent responses have not changed for end-user inputs defined in the test case.
+references: !ruby/object:Api::Resource::ReferenceLinks
+  guides:
+    'Official Documentation': 'https://cloud.google.com/dialogflow/cx/docs'
+  api: 'https://cloud.google.com/dialogflow/cx/docs/reference/rest/v3/projects.locations.agents.testCases'
+timeouts: !ruby/object:Api::Timeouts
+  insert_minutes: 40
+  update_minutes: 40
+custom_code: !ruby/object:Provider::Terraform::CustomCode
+  custom_import: templates/terraform/custom_import/dialogflowcx_test_case.go.erb
```
Comment (melinath) [inline mmv1/products/dialogflowcx/TestCase.yaml:29]:
```suggestion
```

File: mmv1/products/dialogflowcx/TestCase.yaml [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+      The conversation turns uttered when the test case was created, in chronological order. These include the canonical set of agent utterances that should occur when the agent is working properly.
+    item_type: !ruby/object:Api::Type::NestedObject
+      properties:
+        - !ruby/object:Api::Type::NestedObject
+          name: 'userInput'
+          description: |
+            The user input.
+          properties:
+            - !ruby/object:Api::Type::NestedObject
+              name: 'input'
```
Comment (melinath) [inline mmv1/products/dialogflowcx/TestCase.yaml:116]: It looks like `intent` and `audio` are missing here. This is already a pretty big PR so I'm fine with leaving them for later assuming that's intentional.

File: mmv1/products/dialogflowcx/TestCase.yaml [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+                function: 'validation.StringIsJSON'
+            - !ruby/object:Api::Type::Boolean
+              name: 'isWebhookEnabled'
+              description: |
+                If webhooks should be allowed to trigger in response to the user utterance. Often if parameters are injected, webhooks should not be enabled.
+            - !ruby/object:Api::Type::Boolean
+              name: 'enableSentimentAnalysis'
+              description: |
+                Whether sentiment analysis is enabled.
+        - !ruby/object:Api::Type::NestedObject
```
Comment (melinath) [inline mmv1/products/dialogflowcx/TestCase.yaml:180]: it looks like `status` is missing here but again happy to leave that for a separate PR. Or possibly this is only present in the last_test_result?

File: mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  notes        = "demonstrates a simple training phrase response"
+
+  test_config {
+    tracking_parameters = ["some_param"]
+    page                = google_dialogflow_cx_page.page.id
+  }
+
+  test_case_conversation_turns {
+    user_input {
+      input {
```
Comment (melinath) [inline mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go:197]: are these updatable? if so would be great to exercise

File: mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+        language_code = "en"
+        text {
+          text = "training phrase"
+        }
+      }
+      injected_parameters       = jsonencode({ some_param = "1" })
+      is_webhook_enabled        = false
```
Comment (melinath) [inline mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go:204]: `false` and empty are generally considered the same by Terraform - would it be possible to set these to `true` instead?

File: mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    virtual_agent_output {
+      session_parameters = jsonencode({ some_param = "1" })
+      triggered_intent {
+        name         = google_dialogflow_cx_intent.intent.id
+      }
+      current_page {
+        name         = google_dialogflow_cx_page.page.id
+      }
+      text_responses {
+        text = ["Training phrase response"]
```
Comment (melinath) [inline mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go:216]: can this text be changed?

**Review by melinath — CHANGES_REQUESTED** [review]
A few small changes & questions - this is looking really good

File: mmv1/products/dialogflowcx/TestCase.yaml (hunk 1 above)
Comment (mmurakowski-verily) [reply mmv1/products/dialogflowcx/TestCase.yaml]: TIL; in my defense, the other dialogflowcx resources use custom imports, probably because they predate `%parent` 😅.

TODO for myself: make a follow-up cleanup PR to update other dialogflowcx resources to use `%parent` assuming this works.

File: mmv1/products/dialogflowcx/TestCase.yaml (hunk 3 above)
Comment (mmurakowski-verily) [reply mmv1/products/dialogflowcx/TestCase.yaml:115]: Unfortunately, TestCases don't support `intent` or `audio`, per [the documentation](https://cloud.google.com/dialogflow/cx/docs/reference/rest/v3/ConversationTurn#userinput). I've added a comment to this effect.

File: mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go (hunk 7 above)
Comment (mmurakowski-verily) [reply mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go:216]: Done!

File: mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go (hunk 6 above)
Comment (mmurakowski-verily) [reply mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go]: Oops, good point, done.

File: mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go (hunk 5 above)
Comment (mmurakowski-verily) [reply mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go:197]: Done, good point.

File: mmv1/products/dialogflowcx/TestCase.yaml (hunk 4 above)
Comment (mmurakowski-verily) [reply mmv1/products/dialogflowcx/TestCase.yaml:180]: I wasn't able to make it work in test case inputs, I _think_ it's only populated in `last_test_result` but acknowledge that I might just be doing it wrong 😬

Will mark as "resolved" for now.

File: mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go (hunk 7 above)
Comment (mmurakowski-verily) [reply mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go:216]: (I did it in the "basic" test, not here; it's updating the input/response though)

File: mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go (hunk 5 above)
Comment (mmurakowski-verily) [reply mmv1/third_party/terraform/services/dialogflowcx/resource_dialogflow_cx_test_case_test.go:197]: (I did it in the "basic" test, not here; it's updating the input/response though)

**Review by melinath — APPROVED** [review]
LGTM
